ICSharpCode.Decompiler/TypeSystem/Implementation/MetadataField.cs

```diff
@@ -225,7 +225,6 @@ private IType DecodeTypeAndVolatileFlag()
 				if (ty is ModifiedType mod && mod.Modifier.Name == "IsVolatile" && mod.Modifier.Namespace == "System.Runtime.CompilerServices")
 				{
 					Volatile.Write(ref this.isVolatile, true);
-					ty = mod.ElementType;
```

Comment:
We need to get rid of the `ModifiedType`, otherwise it might confuse other parts of the type system that don't expect it to be present (e.g. in overload resolution).
But I think you can re-order the `ApplyAttributesToType` call to come before this if.

Reply:
Doesn't `ApplyAttributesToType` already remove modifier types? Or is that option not enabled by default?

Reply:
Ah right, I forgot that `ApplyAttributesToType` usually removed modopts.
Pretty much only the ILSpy UI uses `TypeSystemOptions.KeepModifiers` to keep them around, and in that case it seems harmless to also keep `IsVolatile` around.